internal/common/tree.go

```diff
@@ -176,84 +239,59 @@ func RemoveLabelFromTree(node *DNSTreeNode, label string) {
 }
 
 func AddLabelToTree(node *DNSTreeNode, label, value string) {
+	if len(node.Endpoints) == 0 {
+		node.Endpoints = append(node.Endpoints, &endpoint.Endpoint{Labels: map[string]string{label: value}})
+	} else {
+		for _, ep := range node.Endpoints {
+			if ep.Labels == nil {
+				ep.Labels = endpoint.NewLabels()
+			}
+			ep.Labels[label] = value
+		}
+	}
 	for _, c := range node.Children {
 		AddLabelToBranch(node, c.Name, label, value)
 		AddLabelToTree(c, label, value)
 	}
 }
 
-func HasLabelForBranch(node *DNSTreeNode, branch, label, value string) bool {
-	d := findDataSetForChild(node, branch)
-	if d == nil {
-		return false
-	}
-	if v, ok := d.Labels[label]; ok {
-		return value == v
+func hasLabelForBranch(node *DNSTreeNode, branch, label, value string) bool {
+	for _, ep := range findEndpointsForChild(node, branch) {
+		if v, ok := ep.Labels[label]; ok {
+			return value == v
+		}
```

Comment:
The "it could be easier" bit (ignorable):
It is probably just me but "branch" means everything under the child, while in this code the "branch" seems to mean "node" :thinking:

Reply:
yes, I agree with the phraseology, if we're saying each child of a node is a branch from that node. 

Then this function is asking "does this node have this label for it's child(branch) with this name".